### DISCUSSION TITLE
Corporate bonds as input

### DISCUSSION THREAD

**Discussion by Conway-Yu** [skfolio/skfolio#60] [Q&A] [answered]
Hello,

I am trying to construct a portfolio using corporate bonds and optimize vs. a benchmark.
The basic idea is to pick from an universe of bonds the best allocation at time t.
When I try to use the framework to fit my X (bonds with total return) to Y (benchmark), I get an error as corporate bonds do not have a continuous time series (contain NaN).
For example a bond starting in 2005 with a Maturity of 10 years won't have data from 2015 to today.

Do I understand correctly that skfolio currently does not support such a portfolio construction?

P.S. I am also working on a pull request right now, as I implemented the market invariant for bonds. So this is shouldn't be an issue when optimizing the portfolio.

**Comment by HugoDelatte** [comment]
You are correct. I think the most elegant solution would be to implement a `Transformer` that drops assets that haven't started yet (via a pre-selection transformer [like this](https://skfolio.org/user_guide/pre_selection.html))
We could also create a `Transformer` that input missing values (an `Imputer`).

Then use them in a `Pipeline` with the optimization.

The pre-selection transformer should be straightforward to implement, after we agreed on the best approach I can add it to the package.

Looking forward for your PR on the market invariant for bonds!

Also, we are working on [this new feature](https://github.com/skfolio/skfolio/issues/56) that could be useful if you need to work with extrageneous time series.

**Reply by Conway-Yu** [reply]
Thanks for the answer.

Yes, I generally agree with the ``Transformer`` approach.

My thoughts were that the best option would be to consider only the available assets at time t.
This would mean to neglect bonds that have not started yet and also the ones that have stopped existing.
``Imputation`` would I think not yield the desired results as I have total return data and e.g. imputing 0 would yield the information that the bond had a zero total return at time t. 

However, as far as my research went is that it is not possible to consider e.g. at time t bonds 1-5 and at time t+10 bonds 4-8 via a transformer approach (or I might be highly mistaken).

So my current approach would probably to iterate through time and use the skfolios framework to fit the available bonds at each time point. However, this would lead to open question for e.g. covariance matrix estimation etc.

Do you have an opinion on that?

**Reply by HugoDelatte** [reply]
Yes I agree that `Imputation` alone is not the right approach for this case (assets that haven't started yet or stopped existing).
Yes, direct iteration without a Pipeline would easily work but if we can find a way to do it with Transformers, then we can leverage the entire toolkit (`cross_val_predict`, `GridSearchCV` etc.)

I think that the `Transformer` approach would work (unless I misunderstood your problem).
I'll break it down to 3 different problems that require 3 different solutions.

## 1) Assets that haven't started yet
Let's assume we have 3 assets (asset 3 starts at time 4):
`r1, r2, r3, r4, r5, r6, r7, r8, r9,`
`r1, r2, r3, r4, r5, r6, r7, r8, r9`
`NaN, NaN, NaN, r4, r5, r6, r7, r8, r9`

The solution is to build a pre-selection Transformer that simply drops assets that contain NaNs.
Let's assume you build a Pipeline with this pre-selection Transformer + Optimization and apply `cross_val_predict` using `WalkForward` that splits the data in 3 periods (1,2,3 & 4,5,6 & 7,8,9):
* The train on period 1 will select assets 1 & 2 (asset 3 is dropped because of NaNs), and apply these weights for the test in period 2.
* The train on period 2 will select assets 1 & 2 & 3, and apply these weights for the test in period 3.

## 2) Assets Expired
In the case of assets that have a deterministic expiration date (like bonds), an investor may want to remove these assets from the optimization ahead of time (because the expiration is known at the beginning, this is not data leakage).

Let's assume we have 3 assets (asset 3 expires at time 7):
`r1, r2, r3, r4, r5, r6, r7, r8, r9,`
`r1, r2, r3, r4, r5, r6, r7, r8, r9`
`r1, r2, r3, r4, r5, r6, r7, NaN, NaN`

The solution is to build a pre-selection Transformer in which you provide as parameter the expiration calendar and drops assets that will expire in `n` observations (or in the next fold).

Let's assume you build a Pipeline with this pre-selection Transformer + Optimization and apply `cross_val_predict` using `WalkForward` that splits the data in 3 periods (1,2,3 & 4,5,6 & 7,8,9):
* The train on period 1 will select assets 1 & 2 & 3, and apply these weights for the test in period 2.
* The train on period 2 will select assets 1 & 2 (asset 3 is dropped because the Transform knows that it will expire in the next period), and apply these weights for the test in period 3.

## 3) Assets Defaulted
In that case you don't want to apply 2) otherwise you introduce data leakage (survival bias in this case).

Let's assume we have 3 assets (asset 3 defaults at time 5 meaning that r5 would be -100% for a stock or a function of the recovery rate for bonds):
`r1, r2, r3, r4, r5, r6, r7, r8, r9,`
`r1, r2, r3, r4, r5, r6, r7, r8, r9`
`r1, r2, r3, r4, r5, NaN, NaN, NaN, NaN`

The solution is to use the same pre-selection Transformer as in 1) plus a Zero Imputer.
Let's assume you build a Pipeline with this pre-selection Transformer + Zeros Imputer + Optimization and apply `cross_val_predict` using `WalkForward` that splits the data in 3 periods (1,2,3 & 4,5,6 & 7,8,9):
* The train on period 1 will select assets 1 & 2 & 3, and apply these weights for the test in period 2 (the Imputer will replace the NaN in time 6 by 0).
* The train on period 2 will select assets 1 & 2 (3 is dropped by the Transformer) and apply these weights for the test in period 3

Let me know if that makes sense.

**Reply by Conway-Yu** [reply]
Thank you for the clarification. I agree with the approach to use the transformer with WalkForward.

Personally, I prefer a combination of approaches 1 and 2.

This is due to the nature of corporate bond data providers. Generally, corporate bond data quality is quite poor. As far as I know, existing sources sometimes do not even cover bonds with maturities of less than a year. This makes it difficult to distinguish whether a bond has defaulted or simply is no longer reported. While factors such as ratings and spread developments can be considered, acquiring this extra data is not always straightforward and can unnecessarily complicate the analysis, especially for a general analysis. This is particularly true if one filters out data by avoiding investments in CCC-rated bonds beforehand.

I've included a proposed code solution below. I was uncertain about the best method for filtering out expirations. For illustration purposes, I considered monthly data in the implementation. A more general solution, independent of the time horizon, might involve using integers for the index. However, this approach would not align with the existing test data provided in skfolio.

I am happy to also issue a pull request, unless you want to implement it yourself.

Moreover, I am not quite sure why I get the following warning: "UserWarning: X has feature names, but SelectAvailableData was fitted without feature names", since I tried to follow the implementation of the other transformers from the [docs](https://skfolio.org/_modules/skfolio/pre_selection/_pre_selection.html#SelectKExtremes).

Please let me know your thoughts.

```python
import numpy as np
import numpy.typing as npt
import sklearn.base as skb
import sklearn.feature_selection as skf
import sklearn.utils.validation as skv
from dateutil.relativedelta import relativedelta


class SelectAvailableData(skf.SelectorMixin, skb.BaseEstimator):
    """Transformer for selecting assets that have not expired.

    This transformer removes assets that are set to expire within a certain number of
    observations, according to a predefined expiration calendar.

    Parameters
    ----------
    test_size : int
        Number of observations to look ahead (test sample) for expiration.
    expiration_dates : list of int or None
        A list where each element corresponds to the expiration date (number of observations)
        for each asset. None indicates no expiration for the asset.

    Attributes
    ----------
    to_keep_ : ndarray of shape (n_assets, )
        Boolean array indicating which assets are remaining.

    n_features_in_ : int
        Number of assets seen during `fit`.

    feature_names_in_ : ndarray of shape (`n_features_in_`,)
        Names of assets seen during `fit`. Defined only when `X`
        has assets names that are all strings.
    """

    to_keep_: np.ndarray

    def __init__(self,  test_size: int, expiration_dates: list[int | None] | None = None):
        self.test_size = test_size
        self.expiration_dates = expiration_dates

    def fit(self, X: npt.ArrayLike, y=None):
        """Run the availability selection transformer and get the appropriate assets.

        Parameters
        ----------
        X : array-like of shape (n_observations, n_assets)
            Price returns of the assets.

        y : Ignored
            Not used, present for API consistency by convention.

        Returns
        -------
        self : SelectAvailableData
            Fitted estimator.
        """
        # Do not self validate since data contains nan
        # X = self._validate_data(X)
        n_timestep = X.index[-1]

        # Initialize to_keep_ with whether the asset has nan's in the training period
        to_keep_ = np.all(~np.isnan(X), axis=0)

        # Remove assets that are set to expire within the next x months in the test set
        if self.expiration_dates is not None:
            for i, expiration_date in enumerate(self.expiration_dates):
                if expiration_date is not None and expiration_date <= (n_timestep + relativedelta(months = self.test_size)):
                    to_keep_[i] = False

        self.to_keep_ = np.array(to_keep_)
        return self

    def _get_support_mask(self):
        skv.check_is_fitted(self, 'to_keep_')
        return self.to_keep_
`
```

**Reply by HugoDelatte** [reply]
Yes that's indeed the idea!
I'll implement it as well as the unit tests, add it to the doc and create some example.
I think the warning is coming from the fact that X has not been validated so the features names (from the DataFrame columns) have not been saved in the class attribute. I'll investigate further.

**Reply by Conway-Yu** [reply]
Okay thank you very much!
Making unit tests and adding everything to the doc with examples is a good point. I will keep it in mind the next time!

I just wanted to mention one other note. Using such a transformer and ``WalkForward``, ``cross_val_predict``, ``GridSearchCV`` etc. still yields an error when calling ``predict``. 

This is due to the test set still containing the assets with nan values. To be more precise, when calling ``predict`` the current skfolio functionality instantiates a [``Portfolio``](https://skfolio.org/_modules/skfolio/portfolio/_portfolio.html#Portfolio), which during its ``init`` raises the following error:
```python
536        if np.any(np.isnan(returns)):
537           raise ValueError("NaN found in `returns`")
```

So I was thinking of two solutions:
1) Change the init of ``Portfolio`` by simply filling 0 for the nan values after the matrix multiplication `returns = weights @ rets.T - total_cost - total_fee`

2) Combine your above mentioned three cases by including an ``Imputer`` too. If I understand correctly, the pipe works sequentially. Thus, one could first select all the existing bonds (my minimal example from above) and then impute the rest of the values with 0. So even if the non existing bonds have a value instead of NaN now, they are still not considered in the optimization process (also have a weight of zero in predict). This could look something like the following:
```python
pipe = Pipeline([
    ("select_available_data", SelectAvailableData(6, first_nan_timestamps)),
    ('imputation', SimpleImputer(strategy='constant', fill_value=0)),
    ("optimization", model)
])
```
Personally I find the second solution more beautiful, **if my logic of the pipe working sequentially is not flawed.**

Looking forward to your implementation!

**Reply by HugoDelatte** [reply, marked as answer]
I agree with you, second implementation seems more elegant.
I'm finishing the metadata-routing merging then I'll implement and merge the above.

**Comment by HugoDelatte** [comment]
FYI, work has started in this branch https://github.com/skfolio/skfolio/tree/feature/issue-60/pre-selection-inception-expiry

**Reply by HugoDelatte** [reply]
Hello @Conway-Yu , the PR is ready: https://github.com/skfolio/skfolio/pull/94
Your feedback/review would be welcome :)
I've mentioned your initial code from this discussion as a derivative work in the python file, let me know if you're happy with that.
 
 Also, I've created a new example, let me know what you think:
 
![example](https://github.com/user-attachments/assets/360f6bb8-b071-4f23-ba1b-a367b43384a0)
